go-controller/pkg/libovsdbops/db_index.go

```diff
@@ -0,0 +1,360 @@
+		panic(fmt.Sprintf("NewDbIndex failed: indexT %v should not be empty", indexT))
+	}
+	for _, key := range indexT.GetIDs() {
+		if _, ok := objectIds[key]; !ok {
+			panic(fmt.Sprintf("NewDbIndex failed: required filed %v is not set in %+v", key, objectIds))
+		}
+	}
+	dbIndex := &DbIndex{
+		IndexT:              indexT,
+		OwnerControllerName: controller,
+		ObjectIDs:           objectIds,
+	}
+	return dbIndex
+}
+
+// GetExternalIDs returns a map of ids, that always includes keys
+// - OwnerController
+// - OwnerObjectType
+// - PrimaryIDKey (constructed by DbIndex.GetUniqueID)
+// SecondaryIndexKey key is included if it is returned by dbIndex.ObjectIDs.GetExternalIDs().
+// Object-specific ids (except SecondaryIndexKey) are returned based on DbIndex.ObjectIDs.GetExternalIDs(),
+// with keys prefixed with "<DbIndex.OwnerObjectType>.<DbIndex.OwnerControllerName>.k8s.ovn.org/".
+// Keys with empty string values will be omitted as unset, e.g. for predicate search.
+func (dbIndex *DbIndex) GetExternalIDs() map[string]string {
+	externalIDs := dbIndex.GetExternalIDsWithoutPrimary()
+	externalIDs[PrimaryIDKey] = dbIndex.GetUniqueID()
+	return externalIDs
+}
+
+// GetUniqueID returns unique string that identifies object with given DbIndex.
+// Different DbIndexes should always have different ids, equal DbIndexes will return the same string.
+// Returned string start with "<dbIndex.OwnerControllerName>:<dbIndex.IndexT.ownerObjectType>"
+// After that values from dbIndex.ObjectIDs are added in order set in dbIndexes, and are separated by ":".
+// For DbIndex(AddressSetEgressFirewallDNS, "controller-1", )
+func (dbIndex *DbIndex) GetUniqueID() string {
+	id := dbIndex.OwnerControllerName + ":" + string(dbIndex.IndexT.ownerObjectType)
+	for _, key := range dbIndex.IndexT.GetIDs() {
+		id += ":" + dbIndex.ObjectIDs[key]
+	}
+	return id
+}
+
+func (dbIndex *DbIndex) GetExternalIDsWithoutPrimary() map[string]string {
+	externalIDs := map[string]string{
+		OwnerController: dbIndex.OwnerControllerName,
+		OwnerObjectType: string(dbIndex.IndexT.ownerObjectType),
+	}
+	for _, key := range dbIndex.IndexT.GetIDs() {
+		externalIDs[key] = dbIndex.ObjectIDs[key]
+	}
+	return externalIDs
+}
+
+// GetDbIndexFromExternalIDs is used to parse object ExternalIDs, it uses objectIDsConstructor that should be
+// able to build DbObjectIDs object based on the keys returned by DbObjectIDs.GetExternalIDs().
+func GetDbIndexFromExternalIDs(indexT IndexType, externalIDs map[string]string) (*DbIndex, error) {
+	if externalIDs[OwnerObjectType] != string(indexT.ownerObjectType) {
+		return nil, fmt.Errorf("")
+	}
+	if externalIDs[OwnerController] == "" {
+		return nil, fmt.Errorf("")
+	}
+	requiredIDs := map[string]string{}
+	for _, key := range indexT.GetIDs() {
+		value, ok := externalIDs[key]
+		if !ok {
+			return nil, fmt.Errorf("")
+		}
+		requiredIDs[key] = value
+	}
+	return NewDbIndex(indexT, externalIDs[OwnerController], requiredIDs), nil
+}
+
+type PartialDbIndex DbIndex
```

Comment:
Why would we need this alias? Couldn't this just use `DbIndex` with a different constructor?